### PR TITLE
PR: Ensure Preferences dialog can only be opened once

### PR DESCRIPTION
**Pull request by goanpeca**
# Description

Preferences dialog, should be opened once. Now we make sure that is the case.

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by Carreau — APPROVED** [review]

File: napari/_qt/qt_main_window.py [hunk 1]
```diff
@@ -443,13 +444,22 @@ def _add_file_menu(self):
...
+            )
+
+            if self._qt_window._preferences_dialog_size:
+                win.resize(self._qt_window._preferences_dialog_size)
 
-        if self._qt_window._preferences_dialog_size:
-            win.resize(self._qt_window._preferences_dialog_size)
+            self._qt_window._preferences_dialog = win
+            win.closed.connect(self._on_preferences_closed)
+            win.show()
```
Comment (Carreau) [inline napari/_qt/qt_main_window.py:458]: Don't know if it's possible, but can we make the existing one be on top/ focus if it already exists ?

**Comment by tlambert03** [comment]
can we use `exec_()` instead of `show()` here?  We are generally trying to be pretty aggressive about minimizing open windows in napari.  This would mean that the user has to close the preferences window before continuing, (but also solves the duplicate window and window-on-top problems naturally).  @jni and @sofroniewn, thoughts on that?

**Comment by goanpeca** [comment]
>can we use exec_() instead of show() here? 

That was my first thought too. I feel like being able to do some stuff and still have the prefs open might be desirable and that is why I changed to this.

But if we agree on having a modal dialog (exec_()) that also works :), and simplifies the PR.

File: napari/_qt/qt_main_window.py (hunk 1 above)
Comment (goanpeca) [reply napari/_qt/qt_main_window.py:458]: >Don't know if it's possible, but can we make the existing one be on top/ focus if it already exists ?

Yes, that is what usually happens in other software that do not have modal preference.

Good catch.

**Comment by sofroniewn** [comment]
> can we use exec_() instead of show() here? We are generally trying to be pretty aggressive about minimizing open windows in napari. This would mean that the user has to close the preferences window before continuing, (but also solves the duplicate window and window-on-top problems naturally). @jni and @sofroniewn, thoughts on that?

indeed we have been, definitely always important to be mindful of window explosion

i am thinking though here that people might want to edit preferences during a session - particularly if we follow through with https://github.com/napari/napari/issues/2451 and move over the plugin ordering dialog.

My vote though would be to turn the preferences dialog into a dock widget to then. I really like being able to dock the windows that I keep open (though one doesn't have to). How does everyone else feel about that? If people like that idea I might suggest we even do it in this PR

**Comment by Carreau** [comment]
> How does everyone else feel about that? If people like that idea I might suggest we even do it in this PR

I would run that through Lia.

I think that panels should be reserved for tools that directly affect the main task of Napari, and don't require restart, confirmation.

IMHO as a panel it might feel weird as the preference window has its own navigation side bar, and values get persisted applied only on "Ok", and reverted in "Cancel". We also expect some values to need restart.

**Comment by goanpeca** [comment]
>How does everyone else feel about that? If people like that idea I might suggest we even do it in this PR

>I would run that through Lia.

Agreed. Also I would be not so inclined to do this for the same reasons @Carreau exposed.


>IMHO as a panel it might feel weird as the preference window has its own navigation side bar, and values get persisted applied only on "Ok", and reverted in "Cancel". We also expect some values to need restart.

Agree with this!

**Comment by tlambert03** [comment]
I think I agree.  Preference windows are so stereotyped, having a dockable one might be too different from other apps.

One of the main reasons we had the plugin sorter be dockable and non-modal was so that someone could "test" it out, opening files while sorting.  But I think maybe a better UI there would be to just let them drop a file somewhere in the sorter itself to see which plugin would claim it?

**Comment by sofroniewn** [comment]
Ok, the reasons against dockable make sense.

My vote is then for a modal dialog (using exec). Especially if it will simplify the code. I just tried out some other desktop software and their preferences are modal, but they will update things for you when you change things like the theme. They also have a "prev" button if we need it, and I like @tlambert03 suggestion that for things like the plugin sorter we could have a little "test area".

How do people feel about going with modal + exec for this PR then? @goanpeca @Carreau @liaprins-czi @tlambert03

**Comment by liaprins-czi** [comment]
I agree with all about the Preferences being _not_ dockable, and keeping as a modal.

**Comment by goanpeca** [comment]
Will make the changes.

**Comment by jni** [comment]
Guys, sorry that I'm late to the party here, and I'm not *too* strongly against modal, but I would prefer if we went for a regular window. I agree with @goanpeca's initial instinct that people expect preferences to be able to be left open while the app is running. It's absolutely the standard on macOS and most software in Linux. I think Windows too? I don't consider this part of the window explosion problem, it's a very specific and very standard window.

**Comment by sofroniewn** [comment]
>  I'm not too strongly against modal, but I would prefer if we went for a regular window .... It's absolutely the standard on macOS and most software in Linux. I think Windows too?

Further investigation has reveled to me that it is at least mixed on macOS - the first programs I tested were photoshop and illustrator which are both model, but then I tried preview and iTerm which are non-modal. Chrome opens in a tab, as does VSCode, which is slightly different yet again, but closer to a non-modal experience - maybe actually closer to a dockable experience ;-) Of those things listed, personally I see us as most similar to photoshop/ illustrator use case which is modal, but that need not be the deciding factor.

It seems like the majority of people have been on the modal, non-dockable side, so unless more people want to speak up or some of those want to change their mind/ voice stronger support non-modal based on this information, I say we stick with modal for now, but I'm certainly open to hearing more points of view on this thread/ more reasons in either direction if there is a strong case to be made.

> I agree with all about the Preferences being not dockable, and keeping as a modal.

Maybe @liaprins-czi could add a quick note expanding on the reason for her modal recommendation? I will also note the right now we are in fact non-modal so maybe there is an argument for just keeping what we have too.

**Comment by jni** [comment]
@sofroniewn all the apps you list as counterarguments are non-native, and have always been so. e.g. Illustrator & co were laggards in the transition to retina displays because they don't use macOS native APIs or patterns. If you look at all the Apple native apps (Finder, Pages, Mail, etc), they use a non-modal panel. In fact, as I suspected there are Apple HCI guidelines for this:

https://developer.apple.com/design/human-interface-guidelines/macos/app-architecture/preferences/

> A preferences window is a modeless dialog that’s used to adjust an app’s preferences.

**Comment by liaprins-czi** [comment]
I actually don't have a strong preference whether it should be a window or modal, more so just that it should not be the dockable sort. I think I misunderstood when I was reading the original comments about changing to a modal, and thought it needed to change to that for technical / implementation reasons. I actually think I slightly prefer it to be able to stay open while using the main window(s) of the viewer, if possible.

**Comment by sofroniewn** [comment]
Ok sounds like there are some compelling arguments in favor of non-dockable non-modal windows now, which I am fine with too.

@goanpeca am I correct that is what is currently implemented here? If so then we can give this a final review. Is there anything else that you want to do to this PR before then? Thanks for bearing with us during all the discussion, but I think it was good to have!

**Comment by goanpeca** [comment]
Its good for review. The original intent prevailed 🤣

**Comment by goanpeca** [comment]
>Don't know if it's possible, but can we make the existing one be on top/ focus if it already exists ?

Fixed @Carreau  :)

**Review by sofroniewn — APPROVED** [review]
> Its good for review. The original intent prevailed 🤣

Lol, shows that sometimes it's better to let a conversation play out then rush to make changes! This looks good to me. I'm approving now, but will leave this open until tomorrow in case anyone else has additional comments on the implementation, I think we're all aligned now on the functionality!

**Review by tlambert03 — APPROVED** [review]
me too!  Thanks @goanpeca

**Pull request merged**